include/irr/asset/CTexturePacker.h

```diff
@@ -216,53 +206,79 @@ class ICPUTexturePacker : protected ITexturePacker
-            const uint32_t pitch = reg.bufferRowLength*texelSz;
-            for (uint32_t i = 0u; i < reg.imageExtent.height; ++i)
-            {
-                memcpy(dst + ((reg.imageOffset.y + i)*m_physAddrTex->getCreationParameters().extent.width + reg.imageOffset.x)*texelSz, src+i*pitch, reg.imageExtent.width*texelSz);
-            }
-        }
-        //pack tail mips into 1 page (if needed)
-        if (offsets->back().hasMipTailAddr())
-        {
-            const core::vector3du32_SIMD texOffset = pageCoords((*offsets).back().getMipTailAddr(), m_pgSzxy);
-            //TODO
-        }
-        //fill page table
-        const VkExtent3D extent = _img->getCreationParameters().extent;
-        for (uint32_t i = 0u; i < offsets->size(); ++i)
+        const uint32_t texelSz = getTexelOrBlockBytesize(m_physAddrTex->getCreationParameters().format);
+        //fill page table and pack present mips into physical addr texture
+        for (uint32_t i = 0u; i < levelCount; ++i)
         {
-            const uint32_t w = std::max(extent.width>>(_subres.baseMipLevel+i),1u);
-            const uint32_t h = std::max(extent.height>>(_subres.baseMipLevel+i),1u);
-            const uint32_t pgSqrSz = core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+            const uint32_t w = (std::max(extent.width>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+            const uint32_t h = (std::max(extent.height>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
 
             uint32_t* const pgTab = reinterpret_cast<uint32_t*>(
                 reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer()) + m_pageTable->getRegions().begin()[i].bufferOffset
                 );
-            const uint32_t pitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
+            const uint32_t pgtPitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
             const uint32_t pgtH = m_pageTable->getRegions().begin()[i].imageExtent.height;
-            const uint32_t offset = (pitch*pgtH)*(*offsets)[i].layer() + (*offsets)[i].y()*pitch + (*offsets)[i].x();
-            for (uint32_t x = 0u; x < pgSqrSz; ++x)
-                for (uint32_t y = 0u; y < pgSqrSz; ++y)
-                    pgTab[offset + y*pitch + x] = (*offsets)[i].addr;
+            const uint32_t offset = (pgtPitch*pgtH)*pgtOffset.layer() + (pgtOffset.y()>>i)*pgtPitch + (pgtOffset.x()>>i);
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    const uint32_t addr = 0u;//TODO alloc with pool alctr
+                    if (i==(levelCount-1u) && levelCount>_subres.levelCount)
+                    {
+                        //TODO alloc another address and store on upper 16 bits of `addr` (for mip-tail)
+                    }
+                    pgTab[offset + y*pgtPitch + x] = addr;
```

Comment:
if `i+m_pgSzxyLog2+1>=_subres.levelCount` then allocate two pages instead of one (obviously make sure you quit the 3 loops as soon as possible, so you don't hit this multiple times)

Reply:
also its more efficient to allocate addresses from the pool up to 1000 thousand at a time (stack for temporary address storage and function input arguments) using `core::address_allocator_traits<PoolAddressAllocatorTYPE>::multi_alloc(m_pageAllocator,...)`

Reply:
still relevant

Reply:
> core::address_allocator_traits<PoolAddressAllocatorTYPE>::multi_alloc

ah, thats how u do multi_alloc

Reply:
yes, you shouldn't be touching an `address_allocator` directly (just like STD C++ lib doesn't touch allocators directly) but call all their methods via the `traits` wrapper XD